### PR TITLE
wgengine: add OpenBSD router DNS update handling

### PR DESCRIPTION
**Pull request by martinbaillie**
Idiomatic OpenBSD DNS continues to be classic `resolv.conf` (no fancy `resolvconf` like Linux or some BSDs).

I have tested both the replace and restore funcs and they do work. However `Close()` on the routers does not appear be called in the event of SIGINT/TERMs. I think I can see some interrupt handling logic that would propagate a shutdown through the engine in the old `relaynode` code, but not in `tailscaled`.

Also, I'm getting this out the door in the continued spirit of "merge early, merge often" but I'm happy to add tests and refine error log/rets of the BSD stuff as the codebase matures and settles on technique.

Some notes on the TODO comments being removed:
- Tested and confirmed routes are indeed cleaned up nicely by the OpenBSD network stack upon interface deletion.
- Netlink-style monitoring through `devd` should now be possible to add thanks to https://github.com/tailscale/tailscale/pull/41.
- Avoiding prematurely optimising for func reuse across the platforms at this juncture based on a comment in the previous PR.

### CODE REVIEW

**Comment by bradfitz** [comment]
Sorry, I landed a PR of mine that then causes merge conflicts here. Can you rebase?

Also, be sure that your commit message title & PR title clarify that this is a BSD change and not some general DNS update.

**Comment by martinbaillie** [comment]
Rebased and fixed commit/PR wording

File: wgengine/router_openbsd.go [hunk 1]
```diff
@@ -5,25 +5,26 @@
...
 	"github.com/tailscale/wireguard-go/tun"
 	"github.com/tailscale/wireguard-go/wgcfg"
+	"tailscale.com/atomicfile"
 	"tailscale.com/types/logger"
 )
 
-// For now this router only supports the userspace WireGuard implementations.
-//
-// There is an experimental kernel version in the works:
+// For now this router only supports the WireGuard userspaces implementations in
```
Comment (bradfitz) [inline wgengine/router_openbsd.go]: typo: userspace (not userspaces)

File: wgengine/router_openbsd.go [hunk 2]
```diff
@@ -5,25 +5,26 @@
...
+// the BSDs. There is an experimental kernel version in the works for OpenBSD:
```
Comment (bradfitz) [inline wgengine/router_openbsd.go]: I would drop "in the BSDs" and just end the sentence with "implementation".  (also singular)

File: wgengine/router_openbsd.go [hunk 3]
```diff
@@ -168,12 +168,94 @@ func (r *openbsdRouter) Close() error {
...
+	buf := new(bytes.Buffer)
+	fmt.Fprintf(buf, "# resolv.conf(5) file generated by tailscale\n")
+	fmt.Fprintf(buf, "# DO NOT EDIT THIS FILE BY HAND -- CHANGES WILL BE OVERWRITTEN\n\n")
+	for _, ns := range servers {
+		fmt.Fprintf(buf, "nameserver %s\n", ns)
+	}
+	if len(domains) > 0 {
+		fmt.Fprintf(buf, "search "+strings.Join(domains, " ")+"\n")
+	}
+	f, err := ioutil.TempFile(filepath.Dir(tsConf), filepath.Base(tsConf)+".*")
```
Comment (bradfitz) [inline wgengine/router_openbsd.go]: change the `f` var to `tf` for "temp file"?

File: wgengine/router_openbsd.go [hunk 4]
```diff
@@ -168,12 +168,94 @@ func (r *openbsdRouter) Close() error {
...
+	if err != nil {
+		return err
+	}
+	f.Close()
```
Comment (bradfitz) [inline wgengine/router_openbsd.go]: I wouldn't depend on the Name method working after Close. It does, but I don't think that's something we on the Go team ever guaranteed.

Let's just snapshot it for clarity:

```go
    tempName := f.Name()
    f.Close()
```

File: wgengine/router_openbsd.go [hunk 5]
```diff
@@ -168,12 +168,94 @@ func (r *openbsdRouter) Close() error {
...
+	if err := atomicfile.WriteFile(f.Name(), buf.Bytes(), 0644); err != nil {
+		return err
+	}
+	os.Chmod(f.Name(), 0644) // ioutil.TempFile creates the file with 0600.
```
Comment (bradfitz) [inline wgengine/router_openbsd.go]: You're only using TempFile to get a random name, though? atomicfile is doing 0644 and renaming the file into place. So the permissions should still be 0644. I don't think this line is doing anything.

**Comment by martinbaillie** [comment]
Thanks for the feedback @bradfitz. I've rebased and fixed up.

File: wgengine/router_openbsd.go [hunk 6]
```diff
@@ -167,12 +167,95 @@ func (r *openbsdRouter) Close() error {
...
+	if _, err := os.Stat(backupConf); err != nil {
+		if os.IsNotExist(err) {
+			return nil // No backup resolv.conf to restore.
+		}
+		return err
+	}
+	if ln, err := os.Readlink(resolvConf); err != nil {
+		return err
+	} else if ln != tsConf {
+		return fmt.Errorf("resolv.conf is not a symlink to %s", tsConf)
```
Comment (bradfitz) [inline wgengine/router_openbsd.go]: include in error message what it does link to? can do this in follow-up change.

**Pull request merged**